File: examples/c/build_features.cpp
// A minimal C++ example: build a feature matrix over a two-symbol universe.
//
// This goes through `wickra_feature_store.hpp`, the C++ hull shipped beside the
// C header, because that hull is what a C++ caller is meant to use: it owns and
// frees the handle, runs the two-call length protocol behind
// `wickra_feature_store_command` for you, and turns a refusal into an exception
// rather than a negative integer that is easy to ignore. Calling the C
// functions directly from C++ works too -- `build_features.c` shows that -- but
// then the hull would be shipped with only `streaming_test.cpp` building it.
#include <exception>
#include <iostream>
#include <string>

#include "wickra_feature_store.hpp"

namespace {
const char *SPEC =
    R"({"universe":["AAA","BBB"],"features":[)"
    R"({"kind":"indicator","name":"Sma","params":[2]},)"
    R"({"kind":"price","field":"close"}],)"
    R"("labels":[{"kind":"forward_return","horizon":1}]})";

const char *CMD =
    R"({"cmd":"build_batch","data":{)"
    R"("AAA":[)"
    R"({"ts":1,"open":10,"high":10,"low":10,"close":10,"volume":1},)"
    R"({"ts":2,"open":11,"high":11,"low":11,"close":11,"volume":1},)"
    R"({"ts":3,"open":12,"high":12,"low":12,"close":12,"volume":1}],)"
    R"("BBB":[)"
    R"({"ts":1,"open":20,"high":20,"low":20,"close":20,"volume":1},)"
    R"({"ts":2,"open":22,"high":22,"low":22,"close":22,"volume":1},)"
    R"({"ts":3,"open":24,"high":24,"low":24,"close":24,"volume":1}]}})";
}  // namespace

int main() {
    try {
        wickra::FeatureStore store(SPEC);
        const std::string matrix = store.command(CMD);

        std::cout << "wickra-feature-store " << wickra::FeatureStore::version() << "\n";
        std::cout << "matrix: " << matrix << "\n";
    } catch (const std::exception &e) {
        std::cerr << "failed: " << e.what() << "\n";
        return 1;
    }
    return 0;
}

File: examples/r/build_features.R
# A runnable R example: build a feature matrix through the binding.
#
#   cargo build -p wickra-feature-store-c --release
#   export WKFS_INC="$PWD/bindings/c/include"
#   export WKFS_LIB="$PWD/target/release"
#   export LD_LIBRARY_PATH="$WKFS_LIB:$LD_LIBRARY_PATH"   # PATH on Windows
#   R CMD INSTALL bindings/r
#   Rscript examples/r/build_features.R

library(wickrafeaturestore)

spec <- paste0(
  '{"universe":["AAA","BBB"],"features":[',
  '{"kind":"indicator","name":"Sma","params":[2]},',
  '{"kind":"price","field":"close"}],',
  '"labels":[{"kind":"forward_return","horizon":1}]}'
)

candle <- function(time, close) {
  paste0(
    '{"time":', time, ',"open":', close, ',"high":', close,
    ',"low":', close, ',"close":', close, ',"volume":1}'
  )
}

store <- wkfeaturestore_new(spec)
cmd <- paste0(
  '{"cmd":"build_batch","data":{',
  '"AAA":[', candle(1, 10), ",", candle(2, 11), ",", candle(3, 12), "],",
  '"BBB":[', candle(1, 20), ",", candle(2, 22), ",", candle(3, 24), "]}}"
)
response <- wkfeaturestore_command(store, cmd)

cat("wickra-feature-store", wkfeaturestore_version(), "\n")
cat(response, "\n")
